tools/clang/lib/Sema/SemaHLSL.cpp

```diff
@@ -12626,26 +12627,49 @@ HLSLWaveSizeAttr *ValidateWaveSizeAttributes(Sema &S, Decl *D,
       A.getAttributeSpellingListIndex());
 
   pAttr->setSpelledArgsCount(A.getNumArgs());
-  int minWave = pAttr->getMin();
-  int maxWave = pAttr->getMax();
-  int prefWave = pAttr->getPreferred();
 
-  // validate attribute arguments.
-  if (!DXIL::IsValidWaveSizeValue(minWave, maxWave, prefWave)) {
-    S.Diag(A.getLoc(), diag::err_hlsl_wavesize_size)
-        << DXIL::kMinWaveSize << DXIL::kMaxWaveSize;
-  }
+  hlsl::DxilWaveSize waveSize(pAttr->getMin(), pAttr->getMax(),
+                              pAttr->getPreferred());
 
-  bool prefInRange =
-      prefWave == 0 ? true : prefWave >= minWave && prefWave <= maxWave;
-  if (!prefInRange) {
-    S.Diag(A.getLoc(), diag::err_hlsl_wavesize_pref_size_out_of_range)
-        << (unsigned)prefWave << (unsigned)minWave << (unsigned)maxWave;
-  }
+  DxilWaveSize::ValidationResult validationResult = waveSize.Validate();
+
+  // WaveSize validation succeeds when not defined, but since we have an
+  // attribute, this means min was zero, which is invalid for min.
+  if (validationResult == DxilWaveSize::ValidationResult::Success &&
+      !waveSize.IsDefined())
+    validationResult = DxilWaveSize::ValidationResult::InvalidMin;
 
-  if (maxWave != 0 && minWave >= maxWave) {
+  // It is invalid to explicitly specify degenerate cases.
+  if (A.getNumArgs() > 1 && waveSize.Max == 0)
+    validationResult = DxilWaveSize::ValidationResult::InvalidMax;
+  else if (A.getNumArgs() > 2 && waveSize.Preferred == 0)
+    validationResult = DxilWaveSize::ValidationResult::InvalidPreferred;
```

Comment:
I believe it is better to keep the HLSL attribute specific rules in Sema, rather than polluting `DxilWaveSize` with these extra rules, only checked when the suggested parameter is supplied, which only occurs when called from Sema.

In any case, this pertains to the open issue #157.  If that is resolved in a different way than currently implemented, these special cases for Max and Preferred would be deleted anyway.

Reply:
That's fine